cpp/include/cudf/io/json.hpp

```diff
@@ -484,6 +484,352 @@ table_with_metadata read_json(
   json_reader_options options,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/** @} */  // end of group
+
+/**
+ * @addtogroup io_writers
+ * @{
+ * @file
+ */
+
+/**
+ *@brief Builder to build options for `writer_json()`.
+ */
+class json_writer_options_builder;
+
+/**
+ * @brief Settings to use for `write_json()`.
+ */
+class json_writer_options {
+  // Specify the sink to use for writer output
+  sink_info _sink;
+  // Set of columns to output
+  table_view _table;
+  // string to use for null entries
+  std::string _na_rep = "";
+  // Indicates whether to output nulls as 'null'
+  bool _include_nulls = false;
+  // Indicates whether to use JSON lines for records format
+  bool _lines = false;
+  // maximum number of rows to write in each chunk (limits memory use)
+  size_type _rows_per_chunk = std::numeric_limits<size_type>::max();
+  // string to use for values != 0 in INT8 types (default 'true')
+  std::string _true_value = std::string{"true"};
+  // string to use for values == 0 in INT8 types (default 'false')
+  std::string _false_value = std::string{"false"};
+  // Names of all columns; if empty, writer will generate column names
+  // std::vector<std::string> _names;
+  table_metadata const* _metadata = nullptr;  // Optional column names
+
+  /**
+   * @brief Constructor from sink and table.
+   *
+   * @param sink The sink used for writer output
+   * @param table Table to be written to output
+   */
+  explicit json_writer_options(sink_info const& sink, table_view const& table)
+    : _sink(sink), _table(table), _rows_per_chunk(table.num_rows())
+  {
+  }
+
+  friend json_writer_options_builder;
+
+ public:
+  /**
+   * @brief Default constructor.
+   *
+   * This has been added since Cython requires a default constructor to create objects on stack.
+   */
+  explicit json_writer_options() = default;
+
+  /**
+   * @brief Create builder to create `json_writer_options`.
+   *
+   * @param sink The sink used for writer output
+   * @param table Table to be written to output
+   *
+   * @return Builder to build json_writer_options
+   */
+  static json_writer_options_builder builder(sink_info const& sink, table_view const& table);
+
+  /**
+   * @brief Returns sink used for writer output.
+   *
+   * @return sink used for writer output
+   */
+  [[nodiscard]] sink_info const& get_sink() const { return _sink; }
+
+  /**
+   * @brief Returns table that would be written to output.
+   *
+   * @return Table that would be written to output
+   */
+  [[nodiscard]] table_view const& get_table() const { return _table; }
+
+  /**
+   * @brief Returns metadata information.
+   *
+   * @return Metadata information
+   */
+  [[nodiscard]] table_metadata const* get_metadata() const { return _metadata; }
```

Comment:
Something like this should work with std::optional
```suggestion
  [[nodiscard]] table_metadata const* get_metadata() const { return _metadata.has_value() ? &_metadata.get() ? nullptr; }
```